x-pack/plugins/canvas/public/state/reducers/elements.js

```diff
@@ -9,112 +9,159 @@ import { assign, push, del, set } from 'object-path-immutable';
+  return set(workpadState, ['pages', pageIndex, location], newNodes);
 }
 
+const trimPosition = ({ left, top, width, height, angle, parent }) => ({
+  left,
+  top,
+  width,
+  height,
+  angle,
+  parent,
+});
+
+const trimElement = ({ id, position, expression, filter }) => ({
+  id,
+  position: trimPosition(position),
+  ...(position.type !== 'group' && { expression }),
+  ...(filter !== void 0 && { filter }),
+});
+
 export const elementsReducer = handleActions(
   {
     // TODO: This takes the entire element, which is not necessary, it could just take the id.
     [actions.setExpression]: (workpadState, { payload }) => {
       const { expression, pageId, elementId } = payload;
-      return assignElementProperties(workpadState, pageId, elementId, { expression });
+      return assignNodeProperties(workpadState, pageId, elementId, { expression });
     },
     [actions.setFilter]: (workpadState, { payload }) => {
       const { filter, pageId, elementId } = payload;
-      return assignElementProperties(workpadState, pageId, elementId, { filter });
+      return assignNodeProperties(workpadState, pageId, elementId, { filter });
     },
     [actions.setMultiplePositions]: (workpadState, { payload }) =>
       payload.repositionedElements.reduce(
         (previousWorkpadState, { position, pageId, elementId }) =>
-          assignElementProperties(previousWorkpadState, pageId, elementId, { position }),
+          assignNodeProperties(previousWorkpadState, pageId, elementId, {
+            position: trimPosition(position),
+          }),
         workpadState
       ),
     [actions.elementLayer]: (workpadState, { payload: { pageId, elementId, movement } }) => {
-      return moveElementLayer(workpadState, pageId, elementId, movement);
+      const location = getLocationFromIds(workpadState, pageId, elementId);
+      return moveNodeLayer(workpadState, pageId, elementId, movement, location);
     },
     [actions.addElement]: (workpadState, { payload: { pageId, element } }) => {
       const pageIndex = getPageIndexById(workpadState, pageId);
       if (pageIndex < 0) {
         return workpadState;
       }
-
-      return push(workpadState, ['pages', pageIndex, 'elements'], element);
+      return push(
+        workpadState,
+        ['pages', pageIndex, getLocation(element.position.type)],
+        trimElement(element)
+      );
     },
     [actions.duplicateElement]: (workpadState, { payload: { pageId, element } }) => {
       const pageIndex = getPageIndexById(workpadState, pageId);
       if (pageIndex < 0) {
         return workpadState;
       }
-
-      return push(workpadState, ['pages', pageIndex, 'elements'], element);
+      return push(
+        workpadState,
+        ['pages', pageIndex, getLocation(element.position.type)],
+        trimElement(element)
+      );
+    },
+    [actions.rawDuplicateElement]: (workpadState, { payload: { pageId, element } }) => {
```

Comment:
Like the action, this is duplicating a lot of code from `duplicateElement` (in fact, I think the two are identical). Getting back to a single action would fix it, otherwise pulling this code into a helper function would be good.

Reply:
Legit point. I have larger concerns as well, which is why I didn't bother, besides of course not getting around to it 😄 For it's quite inefficient to duplicate subtrees element by element. Similar to positioning, I'm planning to switch to bulk subtree duplication and these both will just go away 😄